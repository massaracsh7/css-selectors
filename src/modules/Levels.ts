import { DataLevels, PartsPage } from './types';
import levels from '../assets/levels';
import Reset from './Reset';
import Game from './Game';
import Block from './Block';

export default class Levels extends Game implements PartsPage {
  levelsNew: DataLevels[] = levels;

  viewLevel = (): HTMLElement => {
    const listLevel = document.createElement('ul');
    listLevel.classList.add('level__list');
    for (let i = 0; i < this.levelsNew.length; i += 1) {
      const item = document.createElement('li');
      item.classList.add('level__item');
      item.id = (i).toString();
      if (i === this.levelActive) {
        item.classList.add('level__item--active');
      }
      if (localStorage.results) {
        const levelAnswer = JSON.parse(localStorage.getItem('results') || '');
        if (levelAnswer.right.includes(i)) item.classList.add('level__item--right');
        if (levelAnswer.wrong.includes(i)) item.classList.add('level__item--wrong');
      }
      item.innerHTML = `<span class="level__text">${i + 1}.${this.levelsNew[i].helpTitle}</span>`;
      listLevel.append(item);
      item.addEventListener('click', () => {
        this.levelActive = +item.id;
        this.loadLevel();
        item.classList.add('level__item--active');
      });
    }

    const btn = new Reset();

    const containerLevel = new Block('div', 'level', listLevel, btn.create());
    return containerLevel.create();
  };
}
