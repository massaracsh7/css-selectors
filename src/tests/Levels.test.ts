/** @type {import('<rootDir>/node_modules/ts-jest/dist/types').InitialOptionsTsJest} */

const mockHelpTitle: { [key: string]: string } = {
  'TST0': 'Select by type A',
  'TST1': 'Select by .classname',
  'TST2': 'Combine A.classname',
  'TST3': 'Select by #id'
};

const mockQuestions: { [key: string]: string[] } = {
  'TST0': ['duck', 'zebra', 'giraffe', 'flamingo'],
  'TST1': ['ballRound', 'circleDonnut', 'ballStar', 'flamingo'],
  'TST2': ['zebra', 'circleWatermelon', 'duck', 'circleDonnut'],
  'TST3': ['zebra', 'giraffe', 'flamingo', 'duck']
};

const mockGetTitlesFn = jest.fn().mockReturnValue(mockHelpTitle);
const mockGetQuestionsFn = jest.fn().mockReturnValue(mockQuestions);

jest.mock('../assets/levels', () => ({
  helpTitle: mockGetTitlesFn,
  question: mockGetQuestionsFn,
}));

import Levels from '../modules/Levels';

describe('Levels Module', () => {
  let levels: Levels;

  beforeEach(() => {
    levels = new Levels();
  });

  it('should return a defined levelsList', () => {
    const levelsList = levels.viewLevel();
    expect(levelsList).toBeDefined();
  });

  it('should match the snapshot of levelsList', () => {
    const levelsList = levels.viewLevel();
    expect(levelsList).toMatchSnapshot();
  });

  it('should include correct help titles', () => {
    const helpTitles = mockGetTitlesFn();
    Object.keys(helpTitles).forEach(key => {
      expect(helpTitles[key]).toBe(mockHelpTitle[key]);
    });
  });

  it('should include correct questions', () => {
    const questions = mockGetQuestionsFn();
    Object.keys(questions).forEach(key => {
      expect(questions[key]).toEqual(mockQuestions[key]);
    });
  });
});
