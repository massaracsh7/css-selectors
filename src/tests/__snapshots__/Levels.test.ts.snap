// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Levels Module should match the snapshot of levelsList 1`] = `
<div
  class="level"
>
  <ul
    class="level__list"
  />
  <button
    class="reset"
    type="submit"
  >
    reset
  </button>
</div>
`;
